chromium_src/content/browser/tld_ephemeral_storage.cc

```diff
@@ -0,0 +1,72 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/chromium_src/content/public/browser/tld_ephemeral_storage.h"
+
+#include <map>
+#include "base/no_destructor.h"
+#include "content/browser/dom_storage/dom_storage_context_wrapper.h"
+#include "content/browser/dom_storage/session_storage_namespace_impl.h"
+#include "content/browser/renderer_host/render_view_host_delegate.h"
+#include "content/browser/storage_partition_impl.h"
+#include "content/public/browser/render_view_host.h"
+#include "content/public/browser/session_storage_namespace.h"
+#include "content/public/browser/storage_partition.h"
+#include "content/public/browser/web_contents.h"
+#include "services/network/public/mojom/cookie_manager.mojom.h"
+
+namespace content {
+
+namespace {
+
+using TLDEphemeralStorageMap =
+    std::map<TLDEphemeralStorageKey, base::WeakPtr<TLDEphemeralStorage>>;
+
+// This map allows TLDEphemeralStorage to manage the lifetime of ephemeral
+// storage.
+TLDEphemeralStorageMap& active_tld_storage_areas() {
+  static base::NoDestructor<TLDEphemeralStorageMap> active_storage_areas;
+  return *active_storage_areas.get();
+}
+
+}  // namespace
+
+TLDEphemeralStorage::TLDEphemeralStorage(TLDEphemeralStorageKey key,
+                                         StoragePartition* storage_partition)
+    : key_(key), storage_partition_(storage_partition) {
+  DCHECK(active_tld_storage_areas().find(key) ==
+         active_tld_storage_areas().end());
+  active_tld_storage_areas().emplace(key, weak_factory_.GetWeakPtr());
```

Comment:
I don't quite follow why do we need a weak pointer instead of a raw one?

Reply:
In this case the weak pointer is just a sanity check which catches errors where the TLDEphemeralStorage is dereferenced and deleted.

Reply:
which checks do you have in mind?

Reply:
`.get()` on the weak pointer will return nullptr instead of a garbage value, but I will also add a DCHECK to also produce an assertion failure.

Reply:
using `WeakPtr` in this way sounds unusual to me, but probably fine - could you write a comment describing this?

Reply:
Okay. I'll leave a comment.